lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -0,0 +1,242 @@
+            // we may have to deserialize vtables.  If the serialized tables
+            // are damaged, the pass will crash.
+            //
+            // Though, this has the added bonus of not looking into vtables
+            // outside the current module.  Because we're not doing IPA, let
+            // alone cross-module IPA, this is all well and good.
+            auto *BGC = ClassType.getNominalOrBoundGenericNominal();
+            if (BGC && BGC->getModuleContext() != TargetModule) {
+              continue;
+            }
+
+            auto *F = getTargetClassMethod(M, ClassType, CMI);
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+
+          if (auto *WMI = dyn_cast<WitnessMethodInst>(FAI.getCallee())) {
+            SILFunction *F;
+            SILWitnessTable *WT;
+
+            std::tie(F, WT) = M.lookUpFunctionInWitnessTable(
+                WMI->getConformance(), WMI->getMember());
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+        }
+      }
+      return false;
+    }
+
+    using ExitBlockSet = llvm::SmallPtrSetImpl<SILBasicBlock *>;
+    static bool hasInfinitelyRecursiveApply(SILFunction &Fn,
+                                            SILFunction *TargetFn,
+                                            const ExitBlockSet &ExitBlocks) {
+      enum RecursiveState {
+        FoundRecursion          = 0,
+        FoundPathOutOfFunction  = 1,
+      };
+
+      // Set the entry block into the state map.
+      llvm::DenseMap<SILBasicBlock *, RecursiveState> States;
+      States[Fn.getEntryBlock()] = FoundPathOutOfFunction;
+
+      // Make the processing stack and seed it with the entry block.
+      SmallVector<SILBasicBlock *, 16> Stack;
+      Stack.push_back(Fn.getEntryBlock());
+
+      while (!Stack.empty()) {
+        SILBasicBlock *CurBlock = Stack.pop_back_val();
+        RecursiveState &CurState = States[CurBlock];
+
+        if (CurState == FoundPathOutOfFunction) {
+          // Only change state if the block has a recursive call.
+          if (hasRecursiveCallInPath(*CurBlock, TargetFn)) {
+            CurState = FoundRecursion;
+          } else if (ExitBlocks.count(CurBlock)) {
+            // Found a path to the exit node without a recursive call.
+            return false;
+          }
+        }
+
+        // Loop over successor blocks and add them to the Stack if their state
+        // changes.
+        auto Iter = CurBlock->succ_begin();
+        for (auto EndIter = CurBlock->succ_end(); Iter != EndIter; ++Iter) {
+          if (SILBasicBlock *Succ = *Iter) {
+            if (States[Succ] < CurState) {
+              States[Succ] = CurState;
+              Stack.push_back(Succ);
+            }
+          }
+        }
+      }
+
+      // Bail out if all exit nodes are reachable, and only reachable, through
+      // a recursive call.
+      for (auto *EB : ExitBlocks) {
```

Comment:
Why wouldn't the exit block have already been visited by the loop above, which would have early-returned?